orchagent/switchorch.cpp

```diff
@@ -718,6 +951,70 @@ void SwitchOrch::restartCheckReply(const string &op, const string &data, std::ve
     checkRestartReadyDone();
 }
 
+void SwitchOrch::onSwitchAsicSdkHealthEvent(sai_object_id_t switch_id,
+                                            sai_switch_asic_sdk_health_severity_t severity,
+                                            sai_timespec_t timestamp,
+                                            sai_switch_asic_sdk_health_category_t category,
+                                            sai_switch_health_data_t data,
+                                            const sai_u8_list_t &description)
+{
+    std::vector<swss::FieldValueTuple> values;
+    const string &severity_str = switch_asic_sdk_health_event_severity_reverse_map.at(severity);
+    const string &category_str = switch_asic_sdk_health_event_category_reverse_map.at(category);
+    string description_str;
+    const std::time_t &t = timestamp.tv_sec;
+    stringstream time_ss;
+    time_ss << std::put_time(std::localtime(&t), "%Y-%m-%d %H:%M:%S");
+
+    switch (data.data_type)
+    {
+    case SAI_HEALTH_DATA_TYPE_GENERAL:
+    {
+        vector<uint8_t> description_with_terminator(description.list, description.list + description.count);
+        // Add the terminate character
+        description_with_terminator.push_back(0);
+        description_str = string(reinterpret_cast<char*>(description_with_terminator.data()));
+        // Remove unprintable characters but keep CR and NL
+        description_str.erase(std::remove_if(
+                                  description_str.begin(),
+                                  description_str.end(),
+                                  [](unsigned char x) {
+                                      return (x != 0x0d) && (x != 0x0a) && !std::isprint(x);
+                                  }),
+                              description_str.end());
+        break;
+    }
+    default:
+        SWSS_LOG_ERROR("Unknown data type %d when receiving ASIC/SDK health event", data.data_type);
+        // Do not return. The ASIC/SDK health event will still be recorded but without the description
+        break;
+    }
+
+    event_params_t params = {
+        { "sai_timestamp", time_ss.str() },
+        { "severity", severity_str },
+        { "category", category_str },
+        { "description", description_str }};
+
+    if (0 == gMyAsicName.size())
+    {
+        SWSS_LOG_NOTICE("[%s] ASIC/SDK health event occurred at %s, category %s: %s", severity_str.c_str(), time_ss.str().c_str(), category_str.c_str(), description_str.c_str());
+    }
+    else
+    {
+        SWSS_LOG_NOTICE("[%s] ASIC/SDK health event occurred at %s, asic %s, category %s: %s", severity_str.c_str(), time_ss.str().c_str(), gMyAsicName.c_str(), category_str.c_str(), description_str.c_str());
+        params["asic_name"] = gMyAsicName;
+    }
+
+    values.emplace_back("severity", severity_str);
+    values.emplace_back("category", category_str);
+    values.emplace_back("description", description_str);
+
+    m_asicSdkHealthEventTable->set(time_ss.str(),values);
```

Comment:
could it be duplicate timestamp? I see the timestamp is based on second.

Reply:
I do not think so because we do not expect the event to be a frequent event.
But yesterday some others mentioned there can be very frequent events on BRCM platforms. so perhaps we can reconsider it later.

Reply:
Please reconsider. Even if it is not a frequent event, we'd better avoid such case.

Reply:
I think what we need is to have a precise timestamp to differentiate events. See examples in `show reboot-cause history` and `STATE_DB.REBOOT_CAUSE`.
For now, `second` is good enough unless we belive the event can be very frequent.

Reply:
can SDK guarantee that there won't be two event within the same second?

Reply:
The question here is whether we need to maintain a `lossless` semantics here. I do not think so.
The idea of the feature is that we need information to reflect the status of ASIC/SDK and understand and analyze the issues in case there is an ASIC/SDK failure. 
In case there are too frequent events received, eg. more than one in a second, I believe that's enough for us to understand what happened.

Reply:
Offline discussed this topic. We will keep it as it is for now unless it is very likely to have more than 1 event per second on some platforms.
Keep this item open for now.